Reset file to index failed due to ObjectId handling

gerhardol opened the pull request:
Regression from ObjectId handling in #5087

Fixes #5193

Note: The only other command with special handling for ObjectId.InndexId is GetFileBlobHash(), OK special handling there.

Changes proposed in this pull request:
- Exception when resetting to Index
 
Screenshots before and after (if PR changes UI):
- n/a

What did I do to test the code and ensure quality:
- manual test resetting worktree file

Has been tested on (remove any that don't apply):
- Windows 10

RussKie commented on GitCommands/ArgumentBuilderExtensions.cs:
```diff
@@ -209,7 +209,19 @@ string GetArgument()
+            else if (objectId.IsArtificial)
+            {
+                throw new System.ArgumentException("Unexpected artificial commit in Git command: " + objectId);
+            }
```
Can you please move it as the first check?

```
if (objectId.IsArtificial)
{
       throw new System.ArgumentException("Unexpected artificial commit in Git command: " + objectId);
}

if (objectId == ObjectId.IndexId)
{
    // ..
}
else 
{
    // ..
}
```

RussKie commented on GitCommands/ArgumentBuilderExtensions.cs:
```diff
@@ -209,7 +209,19 @@ string GetArgument()
         /// <param name="objectId">The SHA-1 object ID to add to the builder.</param>
         public static void Add(this ArgumentBuilder builder, ObjectId objectId)
         {
```
Do we need to worry about NRE on `objectId`?

RussKie replied:
Please add tests too for this change

RussKie replied:
If it accepts `null`, the implementation must be mindful of that

drewnoakes commented:
Looking at this a bit more, I wonder if this would be a simpler solution that would also prevent other as yet undiscovered bugs with other callers:

```diff
diff --git a/GitCommands/ArgumentBuilderExtensions.cs b/GitCommands/ArgumentBuilderExtensions.cs
index 78916bb70..e0c326643 100644
--- a/GitCommands/ArgumentBuilderExtensions.cs
+++ b/GitCommands/ArgumentBuilderExtensions.cs
@@ -1,4 +1,5 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using System.ComponentModel;
 using GitCommands.Git;
 using GitUIPluginInterfaces;
@@ -205,11 +206,20 @@ string GetArgument()
         /// <summary>
         /// Adds <paramref name="objectId"/> as a SHA-1 argument.
         /// </summary>
+        /// <remarks>
+        /// If <paramref name="objectId"/> is <c>null</c> then no change is made to the arguments.
+        /// </remarks>
         /// <param name="builder">The <see cref="ArgumentBuilder"/> to add arguments to.</param>
-        /// <param name="objectId">The SHA-1 object ID to add to the builder.</param>
-        public static void Add(this ArgumentBuilder builder, ObjectId objectId)
+        /// <param name="objectId">The SHA-1 object ID to add to the builder, or <c>null</c>.</param>
+        /// <exception cref="ArgumentException"><paramref name="objectId"/> represents an artificial commit.</exception>
+        public static void Add(this ArgumentBuilder builder, [CanBeNull] ObjectId objectId)
         {
-            builder.Add(objectId.ToString());
+            if (objectId?.IsArtificial == true)
+            {
+                throw new ArgumentException("Cannot add artificial ObjectId to git command arguments.");
+            }
+
+            builder.Add(objectId?.ToString());
         }
     }
 }
\ No newline at end of file
```

I made that change locally (without any change to `GitModule.cs`) and it fixed #5193. Actually the only bit that's needed is the `?` before the `ToString` call. The rest is just supportive.

What do you think?

---

As an aside, thanks for directing me to using the diff tab in this way. That's quite handy.

gerhardol commented:
> What do you think?

Agree that this is how it should be done.
(GitModule should maybe have a comment too and a test being added.)

Otherwise accepting IndexId should probably be something explicitly accepted.

RussKie commented:
I would suggest reducing cyclomatic complexity:

```
if (objectId == null) 
    return;

// do the rest
```

drewnoakes commented:
> Otherwise accepting IndexId should probably be something explicitly accepted

I can't think of a time when you'd need to pass an artificial commit ID to a git.exe process. Even if a real object ID collides with an artificial one, GE will likely fail elsewhere first. If you really really needed it you could `ToString` the `ObjectId` and to avoid the check via the overload that accepts strings, which never throws.

gerhardol commented:
Updated according to drew's proposal, added tests

> I can't think of a time when you'd need to pass an artificial commit ID to a git.exe process. 

Sure
What I meant was sending artificial ids to the argument builder and let the argument builder resolve the arguments (like Index -> "").
For now, the proposal is the best

drewnoakes commented:
@RussKie please forgive me for merging this while your review is still showing 'changes requested', but I believe @gerhardol has addressed your request, and that Australia is under shadow of night. I am blocked on this fix as part fixes for bare repos.

If there are outstanding changes you want to see here, I'll make sure they get in. I believe that's not the case though, hence merging.
